tests/checkout/tree.c

```diff
@@ -1479,3 +1479,75 @@ void test_checkout_tree__baseline_is_empty_when_no_index(void)
 	git_reference_free(head);
 }
 
+/* ensure that checking out a file with no changes between the baseline
+ * and the target will maintain changes in the index, but forcing will
+ * update that file
+ */
+void test_checkout_tree__modified_in_index_reverted_in_workdir_stays_modified(void)
+{
+	git_index *index;
+	git_reference *head;
+	git_object *obj;
+	git_status_list *status;
+	git_index_entry modified_entry = {{0}};
+
+	assert_on_branch(g_repo, "master");
+	cl_git_pass(git_repository_index(&index, g_repo));
+	cl_git_pass(git_repository_head(&head, g_repo));
+	cl_git_pass(git_reference_peel(&obj, head, GIT_OBJ_COMMIT));
+
+	cl_git_pass(git_reset(g_repo, obj, GIT_RESET_HARD, NULL));
+
+	cl_git_pass(git_status_list_new(&status, g_repo, NULL));
+	cl_assert_equal_i(0, git_status_list_entrycount(status));
+	git_status_list_free(status);
```

Comment:
It's the third test now where I see this pattern, checking that our repository is actually in a clean state (that is, asserting the branch and then asserting that no modifications are present). Maybe put this into a function/macro? I bet it would be useful to other tests.